experimental/ast/context.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ast
+
+import (
+	"fmt"
+
+	"github.com/bufbuild/protocompile/experimental/report"
+	"github.com/bufbuild/protocompile/internal/arena"
+)
+
+// Context is where all of the book-keeping for the AST of a particular file is kept.
+//
+// Virtually all operations inside of package ast2 involve a Context. However, most of
+// the exported types carry their Context with them, so you don't need to worry about
+// passing it around.
+type Context struct {
+	file *report.IndexedFile
+
+	// Storage for tokens.
+	stream          []tokenImpl
+	syntheticTokens []tokenSynthetic
+
+	// This contains materialized literals for some tokens.
+	//
+	// Not all tokens will have an entry here; only those that have "unusual"
+	// representations. This means the lexer can deal with the complex parsing
+	// logic on our behalf in general, but common cases are re-parsed on-demand.
+	//
+	// All elements of this map are string, uint64, or float64.
```

Comment:
Technically, I recently learned that Go's spec thinks that given `map[K]V`, the `V`s are called "elements" (to the point I saw some old-ish commit in `runtime/map.go` from bradfitz correcting comments about this). But I agree, values is more sensible XD.

Reply:
I know reflection calls this elements, but I think that's because they didn't want to add yet-another method to `reflect.Type` -- `Elem()` gets the element type of a slice, pointer, or channel... but _also_ the value type of a map. I think of "elements" of a map being a K+V pair/entry... but what do I know...